casr/src/analysis.rs

```diff
@@ -0,0 +1,428 @@
+            let input = args[at_index + 3].replace("@@", self.path.to_str().unwrap());
+            args[at_index + 3] = input;
+        }
+        let tool_name = tool.file_name().unwrap().to_str().unwrap();
+        if tool_name.ends_with("casr-python") {
+            args.insert(3, "python3".to_string());
+        }
+        if self.at_index.is_none() {
+            args.insert(0, self.path.to_str().unwrap().to_string());
+            args.insert(0, "--stdin".to_string());
+        }
+        if timeout != 0 {
+            args.insert(0, timeout.to_string());
+            args.insert(0, "-t".to_string());
+        }
+
+        let mut casr_cmd = Command::new(tool);
+        casr_cmd.args(&args);
+        casr_cmd.envs(envs);
+
+        debug!("{:?}", casr_cmd);
+
+        // Get output
+        let casr_output = casr_cmd
+            .output()
+            .with_context(|| format!("Couldn't launch {casr_cmd:?}"))?;
+
+        if !casr_output.status.success() {
+            let err = String::from_utf8_lossy(&casr_output.stderr);
+            if err.contains("Timeout") {
+                let mut timeout_name = self
+                    .path
+                    .file_name()
+                    .unwrap()
+                    .to_os_string()
+                    .into_string()
+                    .unwrap();
+                if let Some(idx) = timeout_name.find('-') {
+                    timeout_name.replace_range(..idx, "timeout");
+                }
+                let timeout_path = output_dir.join("timeout").join(timeout_name);
+                if fs::copy(&self.path, timeout_path).is_err() {
+                    error!("Error occurred while copying the file: {:?}", self.path);
+                }
+            } else if err.contains("Out of memory") {
+                let mut oom_name = self
+                    .path
+                    .file_name()
+                    .unwrap()
+                    .to_os_string()
+                    .into_string()
+                    .unwrap();
+                if let Some(idx) = oom_name.find('-') {
+                    oom_name.replace_range(..idx, "oom");
+                }
+                let oom_path = output_dir.join("oom").join(oom_name);
+                if fs::copy(&self.path, oom_path).is_err() {
+                    error!("Error occurred while copying the file: {:?}", self.path);
+                }
+            } else if err.contains("Program terminated (no crash)") {
+                warn!("{}: No crash on input {}", tool_name, self.path.display());
+            } else {
+                error!("{} for input: {}", err.trim(), self.path.display());
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Perform analysis
+///
+/// # Arguments
+///
+/// `matches` - tool arguments
+///
+/// `crashes` - set of crashes, specified as a CrashInfo structure
+///
+/// `tool` - tool that generates reports
+///
+/// `gdb_argv` - arguments for casr-gdb
+pub fn generate_reports(
```

Comment:
Maybe better naming here. Because we do much more than just generate reports. `handle_crashes`?

Reply:
done